File: audience/src/main/java/br/com/concretesolutions/audience/core/actor/ActorRef.java
package br.com.concretesolutions.audience.core.actor;

import br.com.concretesolutions.audience.core.Director;
import br.com.concretesolutions.audience.core.MessageEvent;
import br.com.concretesolutions.audience.core.script.AssistantScript1;
import br.com.concretesolutions.audience.core.script.AssistantScript2;
import br.com.concretesolutions.audience.core.script.Script1;
import br.com.concretesolutions.audience.core.script.Script2;
import br.com.concretesolutions.audience.core.script.Script3;

public final class ActorRef {

    final String ref;

    ActorRef(String ref) {
        this.ref = ref;
    }

    public <T> MessageBuilder<T> tell(T message) {
        return new MessageBuilder<>(message, this);
    }

    public <T> ActorRef passScript(Class<T> clazz, Script1<T> script) {
        Director.getActorRegistry().storageForRef(ref).passScript(clazz, script);
        return this;
    }

    public <T> ActorRef passScript(Class<T> clazz, Script2<T> script) {
        Director.getActorRegistry().storageForRef(ref).passScript(clazz, script);
        return this;
    }

    public <T> ActorRef passScript(Class<T> clazz, Script3<T> script) {
        Director.getActorRegistry().storageForRef(ref).passScript(clazz, script);
        return this;
    }

    public ActorRef passAssistantScript(String eventRef, AssistantScript1 script) {
        Director.getActorRegistry().storageForRef(ref).passAssistantScript(eventRef, script);
        return this;
    }

    public ActorRef passAssistantScript(String eventRef, AssistantScript2 script) {
        Director.getActorRegistry().storageForRef(ref).passAssistantScript(eventRef, script);
        return this;
    }

    public static class MessageBuilder<T> {

        final ActorRef sender;
        final T message;
        ActorRef receiver;
        boolean isOnStage;

        MessageBuilder(T message, ActorRef sender) {
            this.message = message;
            this.sender = sender;
        }

        public void toActor(Class<?> target) {
            to(Director.actorRef(target));
        }

        public void to(ActorRef receiver) {
            this.receiver = receiver;
            Director.play().send(MessageEvent.create(sender, receiver, message, isOnStage));
        }

        public MessageBuilder onStage() {
            this.isOnStage = true;
            return this;
        }

        public MessageBuilder onStage(boolean isOnStage) {
            this.isOnStage = isOnStage;
            return this;
        }
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("ActorRef{");
        sb.append("ref='").append(ref).append('\'');
        sb.append('}');
        return sb.toString();
    }
}

File: audience-core/src/main/kotlin/io/github/victorolinasc/audience/core/actor/ActorRef.kt
package io.github.victorolinasc.audience.core.actor

import io.github.victorolinasc.audience.core.Director
import io.github.victorolinasc.audience.core.MessageEvent
import io.github.victorolinasc.audience.core.script.AssistantScript0
import io.github.victorolinasc.audience.core.script.AssistantScript1
import io.github.victorolinasc.audience.core.script.AssistantScript2
import io.github.victorolinasc.audience.core.script.Script1
import io.github.victorolinasc.audience.core.script.Script2
import io.github.victorolinasc.audience.core.script.Script3

data class ActorRef internal constructor(internal val ref: String) {

    fun <T> tell(message: T): MessageBuilder<T> {
        return MessageBuilder(message, this)
    }

    fun <T> passScript(clazz: Class<T>, script: Script1<T>): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passScript(clazz, script)
        return this
    }

    fun <T> passScript(clazz: Class<T>, script: Script2<T>): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passScript(clazz, script)
        return this
    }

    fun <T> passScript(clazz: Class<T>, script: Script3<T>): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passScript(clazz, script)
        return this
    }

    fun passAssistantScript(eventRef: String, script: AssistantScript0): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passAssistantScript(eventRef, script)
        return this
    }

    fun passAssistantScript(eventRef: String, script: AssistantScript1): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passAssistantScript(eventRef, script)
        return this
    }

    fun passAssistantScript(eventRef: String, script: AssistantScript2): ActorRef {
        Director.system.actorRegistry.storageForRef(ref).passAssistantScript(eventRef, script)
        return this
    }

    class MessageBuilder<T> internal constructor(internal val message: T, internal val sender: ActorRef) {

        private var isOnStage: Boolean = false

        fun toActor(target: Class<out Actor>) {
            to(Director.actorRef(target))
        }

        fun to(receiver: ActorRef) {
            Director.system.send(MessageEvent.create(sender, receiver, message, isOnStage))
        }

        fun onStage(): MessageBuilder<*> {
            this.isOnStage = true
            return this
        }

        fun onStage(isOnStage: Boolean): MessageBuilder<*> {
            this.isOnStage = isOnStage
            return this
        }
    }
}
